### ISSUE TITLE
`KeyPair` `FromStr` and `Deserialize` impls crash

### ISSUE THREAD

**Issue by elsirion** [rust-bitcoin/rust-secp256k1#491]
When trying to deserialize a `KeyPair` the thread crashes with

```
[libsecp256k1] illegal argument. rustsecp256k1_v0_6_1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx)
```

this can be reproduced using the following test without special feature configurations by running `cargo test  test_broken_keypair_serde`. I originally tested using serde, but could simplify it to `FromStr`.

```rust
    #[test]
    fn test_broken_keypair_serde() {
        let ctx = crate::Secp256k1::new();
        let keypair = KeyPair::new(&ctx, &mut thread_rng());
        let msg = keypair.secret_key().secret_bytes().to_hex();
        dbg!(&msg);
        let parsed_key: KeyPair = msg.parse().unwrap();
        assert_eq!(parsed_key, keypair);
    }
```

I'm not familiar enough with the low-level C API of libsecp to propose a fix keeping the unsafe code but would propose to use the global context instead and feature gate these impls accordingly.

**Comment by apoelstra** [comment]
Yikes. Yeah, these are basically completely broken without the `global-context` feature. We will need to backport this probably several versions back.

I think we should use the global context if available and otherwise have fallback code which constructs a signing context. It will be slow but probably better than not working at all, I think.

**Comment by apoelstra** [comment]
Wow, on further investigation it looks like I merged #313, which introduced this code, even though it had no tests and evidently never worked under any circumstances. All the discussion on that PR was about some unrelated terrible idea. What a process failure.
